modules/library/main/src/main/java/org/geotools/gml/producer/CoordinateFormatter.java

```diff
@@ -0,0 +1,129 @@
+/*
+ *    GeoTools - The Open Source Java GIS Toolkit
+ *    http://geotools.org
+ *
+ *    (C) 2018, Open Source Geospatial Foundation (OSGeo)
+ *
+ *    This library is free software; you can redistribute it and/or
+ *    modify it under the terms of the GNU Lesser General Public
+ *    License as published by the Free Software Foundation;
+ *    version 2.1 of the License.
+ *
+ *    This library is distributed in the hope that it will be useful,
+ *    but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ *    Lesser General Public License for more details.
+ */
+package org.geotools.gml.producer;
+
+import java.text.FieldPosition;
+import java.text.NumberFormat;
+import java.util.Locale;
+
+/**
+ * Formats coordinates with a given number of decimals, using code more efficient than NumberFormat when possible
+ */
+public class CoordinateFormatter {
+
+    /**
+     * The min value at which the decimal notation is used 
+     * (below it, the computerized scientific one is used instead)
+     */
+    private static final double DECIMAL_MIN = Math.pow(10, -3);
+
+    /**
+     * The max value at which the decimal notation is used 
+     * (above it, the computerized scientific one is used instead)
+     */
+    private static final double DECIMAL_MAX = Math.pow(10, 7);
+
+    /** To be used for formatting numbers, uses US locale. */
+    private final NumberFormat coordFormatter = NumberFormat.getInstance(Locale.US);
+
+    private final FieldPosition ZERO = new FieldPosition(0);
+
+    /**
+     * The power of ten used for fast rounding
+     */
+    private final double scale;
+
+    /**
+     * Whether we have to format in plain decimal numbers, or we can use scientific notation
+     */
+    private boolean forcedDecimal;
+
+
+    public CoordinateFormatter(int numDecimals) {
+        coordFormatter.setMaximumFractionDigits(numDecimals);
+        coordFormatter.setGroupingUsed(false);
+        scale = Math.pow(10, numDecimals);
+    }
+
+    /**
+     * Formats a number with the configured number of decimals. For better performance best use 
+     * {@link #format(double, StringBuffer)} against a re-used StringBuffer
+     * @param x
+     * @param sb
+     */
+    public String format(double x) {
+        StringBuffer sb = new StringBuffer();
```

Comment:
Just saw that ``AbstractStringBuilder`` is not public, my bad.

Reply:
It's a lot of work for likely no benefit (synch removal is performed by the JVM once it determines there can not be sharing among threads) and it would trickle up in various places (IntelliJ finds 14 references in GeoTools).

Reply:
Interesting, thanks for the feedback.

Reply:
So I agree no point in removing the usage of ``StringBuffer``.

Reply:
In case it is of interest and you were not aware of lock elision already, see this: https://shipilev.net/jvm-anatomy-park/19-lock-elision/